### PR TITLE
fix: handle mismatched types in newOwners/new_owners to prevent NFT loss

### PR DESCRIPTION
**Pull request by SoarinSkySagar**
## Description
Closes issue #227.

## Tasks:
- [x] Modified the `Request` struct in `Protocol.sol` to have `address` type for `newOwners`
- [x] Created a new function `cairoAddressArrayDeserialize` in `Cairo.sol` to deserialize `newOwners` array, with proper comments above the function, in consistence with the rest of the file.
- [x] Used the newly created `cairoAddressArrayDeserialize` function to deserialize the `address` array in `Protocol.sol`.
- [x] Modified the unit tests for `Protocol.sol` in `Protocol.t.sol` so that it is being tested against the proper `newOwners` type.
- [x] Created the unit test function `test_cairoAddressArrayDeserialize` in `Cairo.t.sol` for the function `cairoAddressArrayDeserialize`.

*This PR was created as part of OnlyDust's ODHack 8.0.*

### CODE REVIEW

**Comment by SoarinSkySagar** [comment]
Hi @ptisserand, I have completed all the required tasks and ensured good coding practices in the commit (Issue #227), please review and merge!

**Comment by ptisserand** [comment]
Hi, thanks for your contribution but there is a compilation issue when trying to run tests.
Could you please fix it?

**Comment by SoarinSkySagar** [comment]
hi @ptisserand, yesterday when I ran the test it was compiling, can you please share the screenshot?

**Comment by ptisserand** [comment]
> hi @ptisserand, yesterday when I ran the test it was compiling, can you please share the screenshot of the error so I can look into it?

Ok, maybe your forgot to commit & push, here is the errors:
```
--> src/TestMessaging.sol

Error (7576): Undeclared identifier.
   --> test/Cairo.t.sol:125:9:
    |
125 |         buf1[0] = 0;
    |         ^^^^

Error (7576): Undeclared identifier.
   --> test/Cairo.t.sol:126:93:
    |
126 |         (address[] memory result1, uint256 newOffset1) = Cairo.cairoAddressArrayDeserialize(buf1, 0);
    |                                                                                             ^^^^

Error (7576): Undeclared identifier.
   --> test/Cairo.t.sol:131:9:
    |
131 |         buf2[0] = 2; 
    |         ^^^^

Error (7576): Undeclared identifier.
   --> test/Cairo.t.sol:132:9:
    |
132 |         buf2[1] = uint160(address(0x1234567890abcdef1234567890abcdef12345678));
    |         ^^^^

Error (7576): Undeclared identifier.
   --> test/Cairo.t.sol:133:9:
    |
133 |         buf2[2] = uint160(address(0xabcdefabcdefabcdefabcdefabcdefabcdefabcd));
    |         ^^^^

Error (7576): Undeclared identifier.
   --> test/Cairo.t.sol:135:93:
    |
135 |         (address[] memory result2, uint256 newOffset2) = Cairo.cairoAddressArrayDeserialize(buf2, 0);
    |                                                                                             ^^^^

Error (7576): Undeclared identifier.
   --> test/Cairo.t.sol:142:9:
    |
142 |         buf3[0] = 1; 
    |         ^^^^

Error (7576): Undeclared identifier.
   --> test/Cairo.t.sol:143:9:
    |
143 |         buf3[1] = uint160(address(0xdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef));
    |         ^^^^

Error (7576): Undeclared identifier.
   --> test/Cairo.t.sol:145:93:
    |
145 |         (address[] memory result3, uint256 newOffset3) = Cairo.cairoAddressArrayDeserialize(buf3, 0);
    |                                                                                             ^^^^

Error: 
Compilation failed
```

**Comment by SoarinSkySagar** [comment]
Hi @ptisserand, can you please check now? Earlier the forge commands were running on the main branch of the forked repo whereas I was working on a new branch (even though I changed the branch to the working one), hence no errors were generated. Had to re-clone the forked repo and the errors were there as you said. Here's the new checklist of tasks completed:
- [x] Created a new function `addressArraySerialize` and its corresponding unit test.
- [x] Fixed some type mismatch errors in the `test_cairoAddressArrayDeserialize` function.
- [x] Checked to see that all tests are passing.

### Screenshot of compilation
image

File: apps/blockchain/ethereum/src/sn/Cairo.sol [hunk 1]
```diff
@@ -313,6 +345,33 @@ library Cairo {
...
+    */
+
+    function cairoAddressArrayDeserialize(
+        uint256[] memory buf,
+        uint256 offset
+    )
+        internal
+        pure
+        returns (address[] memory, uint256)
+    {
```
Comment (ptisserand) [inline apps/blockchain/ethereum/src/sn/Cairo.sol:364]: Please use same pattern as other functions 
```solidity
returns (uint256, address[] memory)
```

File: apps/blockchain/ethereum/test/Cairo.t.sol [hunk 2]
```diff
@@ -118,6 +118,72 @@ contract CairoTest is Test {
...
+    function test_cairoAddressArrayDeserialize() public {
+        uint256[] memory buf1 = new uint256[](1);
+        uint256[] memory buf2 = new uint256[](3);
+        uint256[] memory buf3 = new uint256[](2);
+        buf1[0] = 0;
+        (address[] memory result1, uint256 newOffset1) = Cairo.cairoAddressArrayDeserialize(buf1, 0);
+        assertEq(result1.length, 0);
+        assertEq(newOffset1, 1); 
+
+        uint256;
```
Comment (ptisserand) [inline apps/blockchain/ethereum/test/Cairo.t.sol:132]: Please remove this line

File: apps/blockchain/ethereum/test/Cairo.t.sol [hunk 3]
```diff
@@ -118,6 +118,72 @@ contract CairoTest is Test {
...
+        buf2[1] = uint160(address(0x1234567890AbcdEF1234567890aBcdef12345678));
+        buf2[2] = uint160(address(0xABcdEFABcdEFabcdEfAbCdefabcdeFABcDEFabCD));
+
+        (address[] memory result2, uint256 newOffset2) = Cairo.cairoAddressArrayDeserialize(buf2, 0);
+        assertEq(result2.length, 2);
+        assertEq(result2[0], address(0x1234567890AbcdEF1234567890aBcdef12345678));
+        assertEq(result2[1], address(0xABcdEFABcdEFabcdEfAbCdefabcdeFABcDEFabCD));
+        assertEq(newOffset2, 3);
+
+        uint256;
```
Comment (ptisserand) [inline apps/blockchain/ethereum/test/Cairo.t.sol:143]: Please remove this line

File: apps/blockchain/ethereum/test/Cairo.t.sol [hunk 4]
```diff
@@ -118,6 +118,72 @@ contract CairoTest is Test {
...
+    //
+    function test_addressArraySerialize() public {
+        address[] memory arr1 = new address[](2);
+        arr1[0] = address(0x1234567890AbcdEF1234567890aBcdef12345678);
+        arr1[1] = address(0xABcdEFABcdEFabcdEfAbCdefabcdeFABcDEFabCD);
+
+        uint256[] memory buf1 = new uint256[](3);
+        uint256 offset1 = Cairo.addressArraySerialize(arr1, buf1, 0);
+
+        assertEq(buf1[0], 2);
```
Comment (ptisserand) [inline apps/blockchain/ethereum/test/Cairo.t.sol:163]: Could you please add a check on  `offset1` value

File: apps/blockchain/ethereum/test/Cairo.t.sol [hunk 5]
```diff
@@ -118,6 +118,72 @@ contract CairoTest is Test {
...
+        assertEq(buf1[1], uint160(address(0x1234567890AbcdEF1234567890aBcdef12345678))); 
+        assertEq(buf1[2], uint160(address(0xABcdEFABcdEFabcdEfAbCdefabcdeFABcDEFabCD)));  
+
+        address[] memory arr2 = new address[](1);
+        arr2[0] = address(0xDeaDbeefdEAdbeefdEadbEEFdeadbeEFdEaDbeeF);
+
+        uint256[] memory buf2 = new uint256[](2);
+        uint256 offset2 = Cairo.addressArraySerialize(arr2, buf2, 0);
+
+        assertEq(buf2[0], 1);
```
Comment (ptisserand) [inline apps/blockchain/ethereum/test/Cairo.t.sol:173]: Could you please add a check on `offset2` value

File: apps/blockchain/ethereum/test/Cairo.t.sol [hunk 6]
```diff
@@ -118,6 +118,72 @@ contract CairoTest is Test {
...
+        assertEq(buf2[1], uint160(address(0xDeaDbeefdEAdbeefdEadbEEFdeadbeEFdEaDbeeF))); 
+
+        address[] memory arr3 = new address[](0);
+
+        uint256[] memory buf3 = new uint256[](1);
+
+        uint256 offset3 = Cairo.addressArraySerialize(arr3, buf3, 0);
+
+        assertEq(buf3[0], 0);
```
Comment (ptisserand) [inline apps/blockchain/ethereum/test/Cairo.t.sol:182]: Could you please add a check on `offset3` value

**Review by ptisserand — CHANGES_REQUESTED** [review]
Thanks for your PR, some minor fixes remaining and it should be good to been merged.

**Comment by SoarinSkySagar** [comment]
Hi @ptisserand, hope its okay now and and merge-able :)

**Review by ptisserand — APPROVED** [review]
Thanks for your PR!
Let's get it merged

**Pull request merged**